draft-ietf-httpbis-priority.md

```diff
@@ -501,8 +501,69 @@ not specify the incremental(`i`) parameter.
+decisions. No guidance is provided about how this can or should be done. Factors
+such as implementation choices or deployment environment also play a role. Any
+given connection is likely to have many dynamic permutations. For these reasons,
+there is no unilateral perfect scheduler and this document only provides some
+basic recommendations for implementations. 
+
+Clients cannot depend on particular treatment based on priority signals. Servers
+can use other information to prioritize responses.
+
+It is RECOMMENDED that, when possible, servers respect urgency values, sending
+higher urgency responses before lower urgency responses.
+
+It is RECOMMENDED that, when possible, servers respect incremental values.
+Sending non-incremental responses of the same urgency one-by-one, and sending
+incremental responses of the same urgency in round-robin manner.
+
+It is RECOMMENDED that a server considers request generation order as an input
+to prioritization. Prioritizing concurrent requests at the same urgency level
+based on the Stream ID, which corresponds to the order in which clients make
+requests, ensures that clients can use request ordering to influence response
+order.
+
+The incremental parameter ({{incremental}}) indicates how a client processes
+response bytes as they arrive. Non-incremental resources are only used when all
+of the response payload has been received. Incremental resources are used as
+parts, or chunks, of the response payload are received. Therefore, the timing of
+response data reception at the client, such as the time to early bytes or the
+time to receive the entire payload, plays an important role in perceived
+performance. Timings depend on resource size but this scheme provides no
+explicit guidance about how a server should use size as an input to
+prioritization. Instead, the following examples demonstrate how a server that
+strictly abides the scheduling guidance based on urgency and request generation
+order could find that early requests prevent serving of later requests.
+
+1. At the same urgency level, a non-incremental request for a large resource
+   followed by an incremental request for a small resource.
+2. At the same urgency level, an incremental request of indeterminate length
+   followed by a non-incremental large resource.
+
+It is RECOMMENDED that servers avoid such starvation where possible. The method
+to do so is an implementation decision. For example, a server might
+pre-emptively send responses of a particular incremental type based on other
+information such as content size.
+
+HTTP/2 endpoints can advertise that they using the Extensible Priorities scheme
+instead of the HTTP/2 priority scheme by sending
+SETTINGS_DEPRECATE_HTTP2_PRIORITIES; see {{disabling}}. A server that sends or
+receives this setting SHOULD NOT act on priority signals belonging to the HTTP/2
+scheme. The absence of a client Extensible Priority signal SHOULD be treated
```

Comment:
I'm not sure if this sentence is correct. I think that a server is expected to respect the H2 prioritization scheme unless the client sends  SETTINGS_DEPRECATE_HTTP2_SETTINGS.

I also think that we might want to move the suggestion to {{disabling}}, as it talks about how the client handles the existence (or absence) of the settings parameter? Doing so would be fine, as we refer to that section in the previous sentence.

Reply:
I agree this would fit in {{disabling}}. That section includes the sentence 

> The SETTINGS frame precedes any priority signal sent from a client in HTTP/2, so a server can determine if it should respect the HTTP/2 scheme before building state.

So I intended the new text to build on that. My mental model was that a server might just want to cut out a lot of H2 priorities code, leaving only the bits necessary for parsing. Such a server cannot act on the signal and declares so using the deprecate setting. It's unfortunate if the client wants to continue using the old scheme, but there would not be an interop failure.

I could live with downgrading things to only provide this guidance when a server receives the setting. Would that work for you?

Reply:
> I could live with downgrading things to only provide this guidance when a server receives the setting. Would that work for you?

Thanks. I think my preference goes there. IMO, we do not need to recommend servers degrade performance of legacy HTTP/2 clients. So maybe something like: When receiving SETTINGS_DEPRECATE_HTTP2_PRIORITIES, a server MUST ignore the HTTP/2 PRIORITY frames received on that connection.

Reply:
done in https://github.com/httpwg/http-extensions/pull/1266/commits/f29bd50a4d8f9c5ac75b6ddee82844c6fdaa19fd